source/material_model/steinberger.cc

```diff
@@ -252,277 +252,153 @@ namespace aspect
 
 
     template <int dim>
-    double
+    void
     Steinberger<dim>::
-    specific_heat (const double temperature,
-                   const double pressure,
-                   const std::vector<double> &compositional_fields,
-                   const Point<dim> &) const
+    fill_mass_and_volume_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
+                                    std::vector<std::vector<double>> &mass_fractions,
+                                    std::vector<std::vector<double>> &volume_fractions) const
     {
-      double cp = 0.0;
+      // Resize mass and volume fraction vectors
+      mass_fractions.resize(in.n_evaluation_points(), std::vector<double>(material_lookup.size()));
+      volume_fractions.resize(in.n_evaluation_points(), std::vector<double>(material_lookup.size()));
 
-      if (material_lookup.size() == 1)
-        {
-          cp = material_lookup[0]->specific_heat(temperature,pressure);
-        }
-      else if (material_lookup.size() == compositional_fields.size() + 1)
-        {
-          const double background_cp = material_lookup[0]->specific_heat(temperature,pressure);
-          cp = background_cp;
-          for (unsigned int i = 0; i < compositional_fields.size(); ++i)
-            cp += compositional_fields[i] *
-                  (material_lookup[i+1]->specific_heat(temperature,pressure) - background_cp);
-        }
-      else
+      if (has_background)
         {
-          for (unsigned i = 0; i < material_lookup.size(); ++i)
-            cp += compositional_fields[i] * material_lookup[i]->specific_heat(temperature,pressure);
-        }
-
-      return cp;
-    }
-
-
-
-    template <int dim>
-    double
-    Steinberger<dim>::
-    thermal_conductivity (const double,
-                          const double,
-                          const std::vector<double> &,
-                          const Point<dim> &) const
-    {
-      return thermal_conductivity_value;
-    }
-
-
+          for (unsigned int i=0; i<in.n_evaluation_points(); ++i)
+            {
+              // Use the adiabatic pressure instead of the real one, because of oscillations
+              const double pressure = (this->get_adiabatic_conditions().is_initialized())
+                                      ?
+                                      this->get_adiabatic_conditions().pressure(in.position[i])
+                                      :
+                                      in.pressure[i];
+
+              double summed_volumes = 0.;
```

Comment:
I think you can pull the for-loop, the pressure, and the initialization outside of the if-condition. Same for lines 289-291 at the end.

Reply:
Yup!